tachyon/crypto/hashes/sponge/poseidon/halo2_poseidon.h

```diff
@@ -0,0 +1,51 @@
+#ifndef TACHYON_CRYPTO_HASHES_SPONGE_POSEIDON_HALO2_POSEIDON_H_
+#define TACHYON_CRYPTO_HASHES_SPONGE_POSEIDON_HALO2_POSEIDON_H_
+
+#include <vector>
+
+#include "third_party/gmp/include/gmpxx.h"
+
+#include "tachyon/crypto/hashes/sponge/poseidon/poseidon.h"
+
+namespace tachyon::crypto {
+
+template <typename PrimeFieldTy>
+struct Halo2PoseidonSponge : public PoseidonSponge<PrimeFieldTy> {
+  using F = PrimeFieldTy;
+  using State = typename PoseidonSponge<F>::State;
+
+  Halo2PoseidonSponge() = default;
+  explicit Halo2PoseidonSponge(const PoseidonConfig<F>& config) {
+    this->config = config;
+    this->state = State(config.rate + config.capacity);
+    this->state.elements[0] = F::FromMpzClass(mpz_class(1) << 64);
+  }
+
+  std::vector<F> SqueezeNativeFieldElements(size_t num_elements) {
+    std::vector<F> ret =
+        base::CreateVector(num_elements, []() { return F::Zero(); });
+    this->state[this->state.mode.next_index + 1] = F::One();
```

Comment:
@Insun35 I'm not really sure about this twist. Should I increment the next_index by 1 as well?
> Add the finishing sign of the variable length hashing. Note that this mut
        also apply when absorbing line is empty
It says like this in the halo2.

Reply:
It looks like a flag for the finishing sign. I think we don't have a case for re-absorbing the finished state, so that index seems reasonable to me.

Reply:
```suggestion
        size_t squeeze_index = state.mode.next_index;
        if (squeeze_index == config.rate) {
          squeeze_index = 0;
        }
        this->state[squeeze_index + 1] = F::One();
```